Add wildcard uri filter

hmstepanek opened the pull request:
Searching for a partial url is a common use case. It is particularly
useful for publishers who want to know who has been making annotations
on all their articles. It is also useful for users who simply can't
remember the entire url of an article they have annotated.
This is a partial fix for https://github.com/hypothesis/product-backlog/issues/30.

Add UriCombinedWildcardFilter has two configurations:
 1. seperate_keys=True
    Wildcards are only present in wildcard_uri param values and exact
    matches are present in uri/url param values.
 1. seperate_keys=False
    Wildcards and exact matches are present in uri/url param values.

UriCombinedWildcardFilter igores all wildcard uri's that have wildcards
in the domain.

Remove UriFilter from default modifier list in Search class since
the UriCombinedWildcardFilter will need to replace it in certain
cases and there is no way to selectively remove modifiers. The
current design pattern is if the modifier is used by all searches,
then it is in the default list, since it will not be, drop it from
the default list.

Add both UriFilter and UriCombinedWildcardFilter's to the search `__init__`
module as they will be appended to the Search class modifier list as needed.

Modify all existing code that runs Search to append the UriFilter to the list
of modifiers before running a search.

robertknight commented on h/search/query.py:
```diff
@@ -165,6 +165,33 @@ def __call__(self, search, _):
+
+        uris = set()
+        for query_uri in query_uris:
+            expanded = storage.expand_uri(self.request.db, query_uri)
```
The `query_uri` presumably won't match the URLs stored in the index because it includes wildcard `*` markers?

robertknight commented on h/search/query.py:
```diff
@@ -165,6 +165,33 @@ def __call__(self, search, _):
+class UriWildcardFilter(object):
+
+    """
+    A filter that selects only annotations where the 'site' parameter matches as *site*.
```
"site" parameter?

robertknight commented on h/search/parser.py:
```diff
@@ -16,7 +16,7 @@
 
 # Named fields we support when querying (e.g. `user:luke`)
-named_fields = ['user', 'tag', 'group', 'uri', 'url']
+named_fields = ['user', 'tag', 'group', 'uri', 'url', 'wildcard_uri']
```
It might be convenient to support "site:foo.com" or "domain:foo.com" as a shortcut for "httpx://foo.com/*" (ie. all pages on HTTP or HTTPS foo.com URLs). The former is what Google uses for searching eg. all pages on a particular domain.

robertknight commented on h/search/query.py:
```diff
@@ -165,6 +165,33 @@ def __call__(self, search, _):
+            us = [uri.normalize(u) for u in expanded]
+            uris.update(us)
+
+        return search.query(Q('bool', should=[Q('wildcard', **{"target.scope": u}) for u in uris]))
```
The documentation for [Wildcard Query](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-wildcard-query.html) notes that:

> Note that this query can be slow, as it needs to iterate over many terms. In order to prevent extremely slow wildcard queries, a wildcard term should not start with one of the wildcards * or ?.

I think we probably ought to enforce this restriction. Perhaps require that at least protocol + domain appear before the wildcard?

hmstepanek commented:
A couple thoughts on this/things I encountered: 
 1. Ideally I think it would be best to "or" the uri and wildcard_uri fields but I'm not sure how to do this? Wanted to get your thoughts before I pursued that further. 
 1. If you have a url with a port number and specify a `*` after like `http://localhost:3000*` it will through a very ugly url parser error but I'm not sure if that's really necessary to handle as looking at metabase I don't see anything like that?  
 1. If you specify a ? at the end of the url it gets stripped by uri.normalize so I added special handling for that. Alternatively there may be another method we could use in the url parser but I didn't want to risk it handling normalization differently and things not always matching. 
 1. I removed UriFilter as a default from Search. There are other ways we could handle this like adding the ability to remove individual filters but this seemed to keep in line with the current design (only making it a default filter if all searches use it).

robertknight commented:
> If you have a url with a port number and specify a * after like http://localhost:3000* it will throw a very ugly url parser error 

Does the API return a helpful error message and a 4xx status code if the input is invalid. Whatever happens we should avoid returning 5xx errors for problems with the input.

> Ideally I think it would be best to "or" the uri and wildcard_uri fields but I'm not sure how to do this? 

What Elasticsearch query would you ultimately have to generate to achieve this? I suggest working backwards from that.

> If you specify a ? at the end of the url it gets stripped by uri.normalize so I added special handling for that. 

An issue you've stumbled into here is that `?` is a legitimate and _common_ character within URLs, more so than `*`, since it indicates the start of the query string. How does escaping work in the context of wildcard queries? In other words, how would you search for a literal '?' or '*' if you wanted to?

hmstepanek commented on h/search/query.py:
```diff
@@ -15,6 +20,34 @@
+    return False
+
+
+def _pop_param_values(params, key):
```
This was added here because I would like to refactor all classes in here to use this (if they need to) in a follow up PR.

robertknight commented on h/search/query.py:
```diff
@@ -15,6 +20,34 @@
 OFFSET_MAX = 9800
 
 
+def wildcard_uri_is_valid(wildcard_uri):
```
This function returns false if the URL has no path component at all. As a result, in your follow up PR I encountered an error with `/api/search?wildcard_uri=http://example.org`. A wildcard URI with no actual wildcards in it is perhaps a bit pointless but I don't think it should be an error.

hmstepanek replied:
I have fixed this but now it excepts ? inside the domain (not in the scheme just the netloc). I figure this is ok for now and if it causes slowdown issues we can fix it. I don't want to merge this without full-filing the basic requirements and I feel like this is a basic requirement.

robertknight replied:
You've given this a thumbs up but this function still returns `False` if the URI does not contain any wildcards. Is that intentional?

hmstepanek replied:
I thought you were saying it _should_ return False indicating it _is_ an invalid wildcard_uri which I think it should be? I would expect part of validating a wildcard uri is that it does contain wildcards, otherwise it's not a wildcard uri. That way we aren't handling both cases in one method (not that it should be a problem for elasticsearch results) but I don't think we should be using a wildcard search when there aren't any wildcards plus if the user is using our api incorrectly and passing in non-wildcards into the wildcard_uri it notifies them rather than letting them continue to use something incorrectly. Not to mention if the user selected wildcard_uri and didn't provide a wildcard uri, chances are they probably made a mistake and forgot to add a wildcard into the uri.

robertknight commented on h/search/query.py:
```diff
@@ -242,6 +275,91 @@ def __call__(self, search, params):
+            queries.append(Q("terms", **{'target.scope': uris}))
+        return search.query('bool', should=queries)
+
+    def _normalize_uris(self, query_uris, normalize_method=uri.normalize):
```
In testing I discovered an unfortunate issue with our URI normalization. If the path is present but empty, it is removed from the normalized URI. This means that a wildcard search for `https://example.org/*` will not match an annotation on `https://example.org/` since the latter URL gets normalized to `httpx://example.org` in the search index.

I would suggest not trying to fix that in this PR but leaving it as follow-up work.

hmstepanek replied:
^ I think this is the number 1 use case and I really rather address it right now.

hmstepanek replied:
I agree it is a little unfortunate but my thinking there was this would be handled via user input aka: the user would search for https://example.org/* and https://example.org. The problem is when you want to match uri's like httpx://example.org?page=24 since you aren't allowed to specify *'s in the domain of the uri right now.

ajpeddakotla replied:
@robertknight or @hmstepanek can we track this in another issue. Don't think its a huge issue if we don't fix it now, but we should definitely address this in the future. 
> his means that a wildcard search for https://example.org/* will not match an annotation on https://example.org/ since the latter URL gets normalized to httpx://example.org in the search index.
I would suggest not trying to fix that in this PR but leaving it as follow-up work.

hmstepanek replied:
I have fixed this but as a consequence it excepts `?` inside the domain (not in the scheme just the netloc). I figure this is ok for now and if it causes slowdown issues we can fix it. I don't want to merge this without full-filing the basic requirements and I feel like this is a basic requirement. You should be able to wildcard search for  https://example.org*

hmstepanek replied:
I had a better idea for implementation this morning so current state is: we still don't allow wildcard inside the http://foo.com/ but http://foo.com/* normalizes to http://foo.com* now so it does include _all_ domain uri results now.

hmstepanek commented on h/search/query.py:
```diff
@@ -15,6 +17,51 @@
+def wildcard_uri_is_valid(wildcard_uri):
+    """
+    Return True if uri contains wildcards in appropriate places, return False otherwise.
+
```
I really hate this function. Would a regex be simpler?

hmstepanek replied:
I fixed it and now I'm not as hateful. lol

robertknight commented on h/search/query.py:
```diff
@@ -242,6 +286,95 @@ def __call__(self, search, params):
+            wildcard_uris = [u for u in uris if "*" in u or "?" in u]
+            uris = [u for u in uris if "*" not in u and "?" not in u]
+
+        # Ignore all invalid wildcard uris.
```
This comment says _invalid_ wildcard uris but the list comprehension below selects _valid_ uris.

robertknight commented on h/search/query.py:
```diff
@@ -242,6 +286,95 @@ def __call__(self, search, params):
+        elasticsearch. Thus, escaping wildcard characters is not currently
+        supported.
+        """
+        ending_question_wildcard = ""
```
This variable is named `ending_question_wildcard` but it may be an asterisk or question mark. `trailing_wildcard`?

robertknight commented on tests/h/search/query_test.py:
```diff
@@ -496,6 +496,117 @@ def storage(self, patch):
 
+class TestUriCombinedWildcardFilter():
+
+    @pytest.mark.parametrize('params,expected,separate_keys', [
```
It wasn't immediately clear what `expected` is here. IDs? Indexes? Something else?

robertknight commented on h/search/query.py:
```diff
@@ -15,6 +17,48 @@
+        return False
+    # If uri is in the form of a unique id, the domain will not be parsed correctly but
+    # it is still valid. Check for this special case and return early.
+    if wildcard_uri.startswith("urn:x-pdf:"):
```
Should this logic be applied to all non-HTTP URIs? ie. Should this be `if scheme not in ("http", "https")`?

robertknight commented on h/search/query.py:
```diff
@@ -242,6 +286,95 @@ def __call__(self, search, params):
+        It's possible to have a wildcard at the end of a uri, however
+        uri.normalize strips `?`s from the end of uris and something like
+        http://foo.com/* will not be normalized to http://foo.com* without
+        removing the `*` before normalization. To compinsate for this,
```
s/compinsate/compensate

robertknight commented on h/search/query.py:
```diff
@@ -242,6 +286,95 @@ def __call__(self, search, params):
+        removing the `*` before normalization. To compinsate for this,
+        we check for an ending wildcard and add it back after normalization.
+
+        While it's possible to escape `?` and `*` using `\`, the uri.normalize
```
If you use `\` in a string literal (regular string literal or docstring) and are not intending to escape a character, I think this needs to be a raw string - following one of @seanh's most recent PRs.

robertknight commented on tests/h/search/query_test.py:
```diff
@@ -498,7 +498,7 @@ def storage(self, patch):
 class TestUriCombinedWildcardFilter():
 
-    @pytest.mark.parametrize('params,expected,separate_keys', [
+    @pytest.mark.parametrize('params,expected_ann_idxs,separate_keys', [
```
idxs? Did you mean "ids"?

hmstepanek replied:
I really did mean idxs as in indexes. I guess I'll change it to indexes spelled out since it was still unclear.

robertknight commented on h/search/query.py:
```diff
@@ -242,6 +278,95 @@ def __call__(self, search, params):
+        are performed on uri/url parameters.
+    If separate_keys is False:
+        uri/url parameters are expected to contain exact matches and wildcard matches.
+        Values containing wildcards are intrepted as wildcard searches.
```
s/intrepted/interpreted

robertknight commented on h/search/query.py:
```diff
@@ -242,6 +278,95 @@ def __call__(self, search, params):
+        """Initialize a new UriFilter.
+
+        :param request: the pyramid.request object
+        :param separate_keys: if True will treate wildcard_uri as wildcards and uri/url
```
s/treate/treat
